app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt

```diff
@@ -51,16 +42,16 @@ class LockboxAutofillService(
 
     override fun onFillRequest(request: FillRequest, cancellationSignal: CancellationSignal, callback: FillCallback) {
         val structure = request.fillContexts.last().structure
-        val parsedStructure = parseStructure(structure)
+        val parsedStructure = ParsedStructure(structure)
         val requestingPackage = domainFromPackage(structure.activityComponent.packageName)
 
-        if (requestingPackage == null) {
-            callback.onFailure("unexpected package name structure")
+        if (parsedStructure.passwordId == null && parsedStructure.usernameId == null) {
+            callback.onFailure("couldn't find a username or password field")
```

Comment:
_[I **think** this comment is applicable to all `onFailure` calls]_

According to the [API doc](https://developer.android.com/reference/android/service/autofill/FillCallback.html#onFailure(java.lang.CharSequence)), this method is not intended to be used if heuristics fail (e.g., cannot find fields to fill).  This is really handy for debugging and troubleshooting, though.

Should we file a follow-up to change this to `onSuccess(null)` (following the API doc guidelines) when we're closer to public release?

If not, should this string be localized?

Reply:
You're absolutely right about that; I'd made (but not documented) the decision to leave the `onFailure` calls for now for debugging purposes. I'll file a follow-up to change most of these to `onSuccess(null)` for release.

Reply:
#379